code/datums/actions/build_designator.dm

```diff
@@ -0,0 +1,243 @@
+GLOBAL_LIST_INIT(designator_images_list, list(
+	/obj/structure/barricade/solid = image('icons/obj/structures/barricades/metal.dmi', icon_state = "metal_0"),
+	/obj/structure/barricade/solid/plasteel = image('icons/obj/structures/barricades/plasteel.dmi', icon_state = "new_plasteel_0"),
+	/obj/structure/barricade/folding/metal = image('icons/obj/structures/barricades/metal.dmi', icon_state = "folding_metal_0"),
+	/obj/structure/barricade/folding = image('icons/obj/structures/barricades/plasteel.dmi', icon_state = "plasteel_0"),
+))
+
+///Assoc list of construction types to source material
+GLOBAL_LIST_INIT(designator_types, list (
+	/obj/structure/barricade/solid = /obj/item/stack/sheet/metal,
+	/obj/structure/barricade/solid/plasteel = /obj/item/stack/sheet/plasteel,
+	/obj/structure/barricade/folding/metal = /obj/item/stack/sheet/metal,
+	/obj/structure/barricade/folding = /obj/item/stack/sheet/plasteel,
+))
+
+/datum/action/ability/activable/build_designator
+	name = "Construction Designator"
+	desc = "Place a designator for construction."
+	action_icon_state = "build_designator"
+	action_icon = 'icons/mob/actions.dmi'
+	target_flags = ABILITY_TURF_TARGET
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_ABILITY_PLACE_HOLOGRAM,
+		KEYBINDING_ALTERNATE = COMSIG_ABILITY_SELECT_BUILDTYPE,
+	)
+	///personal hologram designator
+	var/obj/effect/build_hologram/hologram
+	/// timerid before we cleanup the hologram
+	var/cleanup_timer
+	/// how long a hologram lasts without movement
+	var/cleanup_time = 4 SECONDS
+	///The typepath of what we want to construct. Typecast for initial var values
+	var/obj/construct_type
+
+/datum/action/ability/activable/build_designator/Destroy()
+	QDEL_NULL(hologram)
+	delete_timer()
+	return ..()
+
+/datum/action/ability/activable/build_designator/should_show()
+	. = ..()
+	if(!.)
+		return
+	return owner.skills.getRating(SKILL_LEADERSHIP) >= SKILL_LEAD_TRAINED
+
+/datum/action/ability/activable/build_designator/can_use_action()
+	return owner.skills.getRating(SKILL_LEADERSHIP) >= SKILL_LEAD_TRAINED
+
+/datum/action/ability/activable/build_designator/fail_activate()
+	if(owner)
+		owner << span_warning("You are not competent enough to do that.") // This message shouldn't show since incompetent people shouldn't have the button, but JIC.
+
+/datum/action/ability/activable/build_designator/on_selection()
+	RegisterSignal(owner, COMSIG_ATOM_MOUSE_ENTERED, PROC_REF(show_hologram_call))
+	RegisterSignal(owner, COMSIG_ATOM_DIR_CHANGE, PROC_REF(on_owner_rotate))
+	RegisterSignal(owner, COMSIG_DO_OVERWATCH_RADIAL, PROC_REF(override_cic_radial))
+
+/datum/action/ability/activable/build_designator/on_deselection()
+	UnregisterSignal(owner, list(COMSIG_ATOM_MOUSE_ENTERED, COMSIG_ATOM_DIR_CHANGE, COMSIG_DO_OVERWATCH_RADIAL))
+	cleanup_hologram()
+
+/datum/action/ability/activable/build_designator/use_ability(atom/A)
+	if(!isturf(A) || !update_hologram(A))
+		owner.balloon_alert(owner, "Invalid spot")
+		return FALSE
```

Comment:
shouldnt this get_turf A first or something? so it works on weeds

Reply:
It uses ABILITY_TURF_TARGET so it should always return the turf unless you're clicking somewhere without one somehow.